### ISSUE TITLE
[MPS] Inconsistent overflow behavior when casting Float to Int8 compared to CPU/CUDA (Saturation vs Wraparound)

### ISSUE THREAD

**Issue by lingebeng** [pytorch/pytorch#169058]
### 🐛 Describe the bug

MPS exhibits implicit saturating cast behavior for floating-point inputs, contradicting the wraparound/truncation behavior observed on CPU and Integer inputs.

### Case 1: Float32 to Int8 (INCONSISTENT)
```python
import torch

x = torch.tensor([128],dtype=torch.float32)
x_mps = x.to("mps")

x = x.to(torch.int8)
x_mps = x_mps.to(torch.int8)

print(x)
print(x_mps)
```
output:
```bash
tensor([-128], dtype=torch.int8) 
tensor([127], device='mps:0', dtype=torch.int8)
```

### Case 2: Int32 to Int8 (CONSISTENT)
```python
import torch
x = torch.tensor([128],dtype=torch.int32)
x_mps = x.to("mps")

x = x.to(torch.int8)
x_mps = x_mps.to(torch.int8)

print(x)
print(x_mps)
```
output:
```bash
tensor([-128], dtype=torch.int8) 
tensor([-128], device='mps:0', dtype=torch.int8)
```

### Case 3: CPU/CUDA (CONSISTENT)
```python
import torch

x = torch.tensor([128],dtype=torch.float32)
x_cuda = x.to("cuda")

x = x.to(torch.int8)
x_cuda = x_cuda.to(torch.int8)

print(x)
print(x_cuda)

print()
x = torch.tensor([128],dtype=torch.int32)
x_cuda = x.to("cuda")

x = x.to(torch.int8)
x_cuda = x_cuda.to(torch.int8)

print(x)
print(x_cuda)
```
output:
```bash
tensor([-128], dtype=torch.int8) 
tensor([-128], device='cuda:0', dtype=torch.int8)

tensor([-128], dtype=torch.int8) 
tensor([-128], device='cuda:0', dtype=torch.int8)
```
### My opinion

I think this inconsistency is very risky.This inconsistency poses a significant threat to quantization workflows.What's more,a model debugged on a Mac (MPS) might hide numerical overflow issues that will later cause unexpected failures or gradient explosions when deployed to a Linux server (CUDA/CPU).I believe to(dtype) should behave identically regardless of the device.

### Versions

PyTorch version: 2.10.0.dev20251119
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.4.4.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.12 (main, Oct 28 2025, 11:52:25) [Clang 20.1.4 ] (64-bit runtime)
Python platform: macOS-26.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M4

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect

**Comment by lingebeng** [comment]
@malfet 
I am so sorry,no docs!
Just for multiple tests

for cpu
128     -> -128
129     -> -127
129.5  -> -127
129.9 -> -127
130    -> -126

for mps
always 127

**Comment by BenjaminDEMAILLE** [comment]
Still reproduces on torch 2.14.0, and the divergence is wider than saturation versus
non-saturation:

```python
x = torch.tensor([-200.0, -129.0, -128.0, 127.0, 128.0, 300.0, float("nan")], device="mps")
x.to(torch.int8)        # mps: [-128, -128, -128, 127, 127, 127, 0]
x.cpu().to(torch.int8)  # cpu: [  56,  127, -128, 127, -128,  44, 0]
```

MPS clamps to the representable range. CPU wraps, which is the C++ float-to-integer
conversion being undefined outside the destination range and the compiler emitting
whatever the target does. `-200 -> 56` and `300 -> 44` are that wrap; `128 -> -128` is
the same thing at the boundary.

Worth separating two questions, because they have different answers:

1. Should MPS match CPU? Matching bit-for-bit means reproducing behaviour that is
   undefined in the C++ sense and varies by host architecture, which does not seem
   like something to spread to another backend.
2. Should the two agree at all? Currently a cast that is fine on one device silently
   produces different numbers on the other, with no warning, which is the part that
   actually bites users.

If the answer to 2 is yes and to 1 is no, then the fix is on the CPU side, or a
documented saturating rule for both, and either is a decision I would rather have from
a maintainer before writing a patch. Happy to implement whichever direction is chosen.

---

*Written with the help of an AI assistant. I ran the repro above myself.*
